### PR TITLE
models: exists function addition

### PR DESCRIPTION
**Pull request by hachreak**

### CODE REVIEW

**Comment by hachreak** [comment]
:shipit: ?

**Comment by jirikuncar** [comment]
@hachreak the `exists` method should not be defined on model. If you want to make it as an API please move it to `invenio_records.api:Record.exist`.

I can't merge the branch because `Coverage decreased (-0.2%) to 25.862%`.

File: invenio_records/api.py [hunk 1]
```diff
@@ -126,13 +126,20 @@ def commit(self):
...
-            obj = RecordMetadata.query.get(recid)
+            obj = models.RecordMetadata.query.get(recid)
         return cls(obj.json, model=obj) if obj else None
 
     def dumps(self, **kwargs):
         # FIXME add keywords filtering
         return dict(self)
 
+    @classmethod
+    def exist(cls, *criteria, **filters):
```
Comment (jirikuncar) [inline invenio_records/api.py]: IMHO we should support only `recid`.

**Comment by jirikuncar** [comment]
@hachreak I'm not sure it's good idea to have `exists` in API. What is the use case?

**Anti-pattern**
```python
if Record.exists(1):
    record = Record.get_record(1)
```

**IMHO Better**
```python
try:
    record = Record.get_record(1)
except sqlalchemy.NoResultFound:
    current_app.logger.exception("Somebody is trying to access record that doesn't exist")
```

--
PS: Ignore PEP8 warnings. GitHub doesn't have linter build in ;)

**Comment by egabancho** [comment]
@jirikuncar Actually the `try` is not even needed as it this exception won't be raised :stuck_out_tongue_winking_eye: Instead get_record will return `None` is there is no record with the given record id

**Comment by kaplun** [comment]
But what if one wants to simply check for existence, without the need of loading a record. Isn't using `get_record` too heavy for that?

**Comment by egabancho** [comment]
@kaplun how would you do it? `get_record` does one query to one table using the primary key ...

**Comment by kaplun** [comment]
Yes but then it returns the whole mega JSON isn't it? While you just want a True/False value. I am thinking about the typical use case of our domain where records are 1 MB big (and need to be parsed into SmartDict etc.)

**Comment by jirikuncar** [comment]
> But what if one wants to simply check for existence, without the need of loading a record.

Present an use-case where you need only True/False and no metadata otherwise it's just nice-to-have.

**Comment by egabancho** [comment]
I see your point and I think it is a valid one, if you don't want to do stuff with the record afterwards, something like (probably a stupid example but it should make the point):
```python
if Record.exits(1):
    extract_references_record(recid=1)
else:
    raise Exception('What I am expected to do here?')
```
I guess we can compare somehow `Record.exits(1)` with `'foo' in dict`.

@jirikuncar WDYT?

**Comment by tiborsimko** [comment]
In `maint-1.2`, there are quite a few use cases for `record_exists()`:

```
$ git grep 'record_exists(' | wc -l
47
```

The feature is used among others to check "soft-deleted" records (when the collection identifier is DELETED). 

- https://github.com/inveniosoftware/invenio/blob/maint-1.2/modules/websearch/lib/search_engine_utils.py#L184-L188

**Comment by jirikuncar** [comment]
1. one wants to know if the `recid` is in the system - PIDStore might be a better place.
2. one is going to work with record metadata, hence use `get_record`.
3. one wants to do a "batch" check, better use model directly.

**Comment by jirikuncar** [comment]
If you still insist on having `exists` method, then there should be "hooks/signals" so modules can enhance the behavior (e.g. `if 'DELETED' in self['_collections']: return False`).

**Comment by kaplun** [comment]
> If you still insist on having exists method, then there should be "hooks/signals" so modules can enhance the behavior (e.g. if 'DELETED' in self['_collections']: return False).

Besides precisely this use case show that you would anyway need to retrieve the whole JSON. So after all it's true that `get_record` is enough as anyway you always to do a minimal check on the metadata.

I was thinking an other use case is the `ListIdentifiers` verb in OAI-PMH, but there too, beside checking for record existence you need to load it anyway to retrieve the OAI-PMH ID (unless this is stored in pidstore).

**Comment by tiborsimko** [comment]
> 1. one wants to know if the recid is in the system - PIDStore might be a better place.

That's a perfectly possible alternative direction, of course.  But in this case the plan:

```
$ grep pidstore setup.py
 'invenio-pidstore>=0.1.1',  # TODO consider making it optional
```

should be revised, because then every record ID should be living in the PID store, together with the status (active, deleted, merged, etc).

**Comment by lnielsen** [comment]
@kaplun @tiborsimko @jirikuncar  I this this PR can be closed after the refactoring of Invenio-Records and PIDStore?

**Comment by kaplun** [comment]
:+1: for me

**Pull request closed without merging**